node/core/provisioner/src/disputes/mod.rs

```diff
@@ -0,0 +1,454 @@
+		}
+	}
+
+	fn into_iter(self) -> impl Iterator<Item = (SessionIndex, CandidateHash)> {
+		self.cant_conclude_onchain
+			.into_iter()
+			.chain(self.unknown_onchain.into_iter())
+			.chain(self.can_conclude_onchain.into_iter())
+			// .chain(self.concluded_known_onchain.into_iter()) -> dropped on purpose
+			.chain(self.concluded_unknown_onchain.into_iter())
+	}
+}
+
+fn partition_recent_disputes(
+	recent: Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	onchain: &HashMap<(u32, CandidateHash), DisputeState>,
+) -> PartitionedDisputes {
+	let mut partitioned = PartitionedDisputes::new();
+
+	// Drop any duplicates
+	let unique_recent = recent
+		.into_iter()
+		.map(|(session_index, candidate_hash, dispute_state)| {
+			((session_index, candidate_hash), dispute_state)
+		})
+		.collect::<HashMap<_, _>>();
+
+	// Split ACTIVE from CONCLUDED disputes
+	let (active, concluded): (
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	) = unique_recent
+		.into_iter()
+		.map(|((session_index, candidate_hash), dispute_state)| {
+			(session_index, candidate_hash, dispute_state)
+		})
+		.partition(|(_, _, status)| *status == DisputeStatus::Active);
+
+	// Split ACTIVE in three groups...
+	for (session_index, candidate_hash, _) in active {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(d) => {
+				// handle double voting here - if a validator has voted both `for` and `against` at the same time for a dispute - ignore its vote
+				// XOR-ing `for` and `against` gives exactly that
+				if (d.validators_for.clone() ^ &d.validators_against).count_ones() >=
+					supermajority_threshold(d.validators_for.len())
+				{
+					partitioned.can_conclude_onchain.push((session_index, candidate_hash));
+				} else {
+					partitioned.cant_conclude_onchain.push((session_index, candidate_hash));
+				}
+			},
+			None => partitioned.unknown_onchain.push((session_index, candidate_hash)),
+		};
+	}
+
+	// TODO: Prioritise the votes in each partition. Is it still necessary?
+
+	// ... and CONCLUDED in two more
+	for (session_index, candidate_hash, _) in concluded {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(_) => partitioned.concluded_known_onchain.push((session_index, candidate_hash)),
+			None => partitioned.concluded_unknown_onchain.push((session_index, candidate_hash)),
+		}
+	}
+
+	partitioned
+}
+
+// Helper trait to obtain the value of vote for `InvalidDisputeStatementKind` and `ValidDisputeStatementKind`.
+// The alternative was to pass a bool to `fn should_keep_vote` explicitly but it's pointless as the value is already 'encoded' in the type.
+trait VoteType {
+	fn vote_value() -> bool;
+}
+
+impl VoteType for InvalidDisputeStatementKind {
+	fn vote_value() -> bool {
+		false
+	}
+}
+
+impl VoteType for ValidDisputeStatementKind {
+	fn vote_value() -> bool {
+		true
+	}
+}
+
+fn should_keep_vote<T: VoteType>(
+	v: &(T, ValidatorIndex, ValidatorSignature),
```

Comment:
Use a `struct VoteWithSignature<T>` rather than a tuple

Reply:
I can't find such struct so I assume you suggest to create a new one and use it instead of the tuple. Right?

If yes - I don't quite understand the whole idea. Won't it create too much boilerplate code - a new struct + conversion from `ValidDisputeStatemientKind` and `InvalidDisputeStatementKind` just for this usecase?